repl/src/main/scala/ammonite/repl/interp/Evaluator.scala

```diff
@@ -206,6 +222,74 @@ object Evaluator{
       )
     }
 
+    def compileCacheBlock(wrapperName: String, code: String, scriptImports: Seq[ImportData]): Res[(Class[_], Seq[ImportData])] = for {
+      _ <- Catching{ case e: ThreadDeath => interrupted() }
+      (classFiles, importData) <- compileClass(
+        s"""
+        $defaultImportBlock
+        ${importBlock(scriptImports )}
+
+        case object $wrapperName{
+
+          $code
+
+          def $$main() = {}
+        }
+        """
+      )
+      
+      _ = cacheSave(wrapperName,(classFiles,importData))
+      cls <- loadClass(wrapperName,classFiles) 
+    } yield (cls, importData)
+
+    def processScriptBlock(code: String, scriptImports: Seq[ImportData]) = {
+      val wrapperName = cacheTag(code, scriptImports)
+      loadCachedClass(wrapperName) match {
+        case Some((cls,importData)) => {
+          evalMain(cls)
+          Res.Success(
+            Evaluated(
+              wrapperName,
+              importData.map(id => id.copy(
+                wrapperName = wrapperName,
+                prefix = if (id.prefix == "") wrapperName else id.prefix
+              ))
+            )
+          )
+        }
+        case None => {
+          for {
+            (cls, newImports) <- compileCacheBlock(wrapperName, code, scriptImports)
+          } yield {
+            evalMain(cls)
+            Evaluated(
+              wrapperName,
+              newImports.map(id => id.copy(
+                wrapperName = wrapperName,
+                prefix = if (id.prefix == "") wrapperName else id.prefix
+              ))
+            )
+          }
+        }
+      }
+    }
+
+    def cacheTag(code: String, imports: Seq[ImportData]): String = {
```

Comment:
Shouldn't the cacheTag also depend on the hash of the classpath when the code was compiled? The same code with the same imports could have wildly different compilations depending both on previous commands and on the jars we had pre-loaded